app/models/concerns/attribute_blockable.rb

```diff
@@ -0,0 +1,355 @@
+    #
+    # @example Basic usage
+    #   attr_blockable :email
+    #   # user.blocked_by_email?
+    #   # user.blocked_by_email_object&.blocked_at
+    #   # user.block_by_email!(by_user_id: current_user.id)
+    #
+    # @example With custom attribute mapping
+    #   attr_blockable :form_email, object_type: :email
+    #   # Uses 'email' BlockedObject type but creates form_email methods
+    #
+    # @example Blocking with expiration
+    #   user.block_by_email!(
+    #     by_user_id: admin.id,
+    #     expires_in: 30.days
+    #   )
+    def attr_blockable(blockable_method, object_type: nil)
+      object_type ||= blockable_method
+
+      define_method("blocked_by_#{blockable_method}?") { blocked_object_by_method(object_type, blockable_method:)&.blocked? || false }
+
+      define_method("blocked_by_#{blockable_method}_object") do
+        blocked_object_by_method(object_type, blockable_method:)
+      end
+
+      define_method("block_by_#{blockable_method}!") do |by_user_id: nil, expires_in: nil|
+        return if (value = send(blockable_method)).blank?
+        block_by_method!(object_type, value, by_user_id:, expires_in:)
+      end
+
+      define_method("unblock_by_#{blockable_method}!") do
+        return if (value = send(blockable_method)).blank?
+        unblock_by_method!(object_type, value)
+      end
+
+      # Register this blockable attribute for introspection
+      self.blockable_attributes = blockable_attributes + [{ object_type: object_type.to_sym, blockable_method: blockable_method.to_sym }]
+    end
+
+    # Returns an array of all blockable method names defined on this model.
+    # Useful for automatically preloading all blockable attributes.
+    #
+    # @return [Array<Symbol>] Array of blockable method names
+    #
+    # @example
+    #   User.blockable_method_names
+    #   # => [:email, :form_email, :form_email_domain]
+    def blockable_method_names
+      blockable_attributes.map { |attr| attr[:blockable_method] }
+    end
+
+    # Preloads all registered blockable attributes for this model.
+    # Convenience method that automatically calls +with_blocked_attributes_for+
+    # with all blockable methods defined via +attr_blockable+.
+    #
+    # @return [ActiveRecord::Relation] Relation with all blockable attributes preloaded
+    #
+    # @example
+    #   User.with_all_blocked_attributes
+    #   # Equivalent to: User.with_blocked_attributes_for(:email, :form_email, :form_email_domain)
+    def with_all_blocked_attributes
+      with_blocked_attributes_for(*blockable_method_names)
+    end
+
+    # Class-level version of +with_blocked_attributes_for+ for use on model classes.
+    #
+    # @param method_names [Array<Symbol, String>] Names of blockable attributes to preload
+    # @return [ActiveRecord::Relation] Relation with preloading configured
+    #
+    # @example
+    #   User.with_blocked_attributes_for(:email, :email_domain)
+    def with_blocked_attributes_for(*method_names)
```

Comment:
Used in
- `Admin::AffiliatesController#index`
- `Admin::SearchController#users`
- `Admin::UsersController#refund_queue`